File: apps/frontend/src/types/serving.ts
/**
 * Model Serving Types
 *
 * Types for model inference service management, A/B testing, and canary deployments.
 */

// ============================================================================
// Enums
// ============================================================================

export enum ServingPlatform {
  KSERVE = 'kserve',
  SELDON = 'seldon',
  TRITON = 'triton',
  CUSTOM = 'custom',
}

export enum ServingStatus {
  PENDING = 'pending',
  DEPLOYING = 'deploying',
  RUNNING = 'running',
  UPDATING = 'updating',
  FAILED = 'failed',
  STOPPED = 'stopped',
  UNKNOWN = 'unknown',
}

export enum PredictorType {
  SKLEARN = 'sklearn',
  XGBOOST = 'xgboost',
  LIGHTGBM = 'lightgbm',
  PYTORCH = 'pytorch',
  TENSORFLOW = 'tensorflow',
  ONNX = 'onnx',
  HUGGINGFACE = 'huggingface',
  CUSTOM = 'custom',
}

export enum DeploymentMode {
  RAW = 'raw',
  AB_TESTING = 'ab_testing',
  CANARY = 'canary',
  SHADOW = 'shadow',
  MIRRORED = 'mirrored',
}

export enum TrafficSplitMethod {
  FIXED = 'fixed',
  EPSILON_GREEDY = 'epsilon_greedy',
  THOMPSON_SAMPLING = 'thompson_sampling',
  UCB1 = 'ucb1',
}

export enum SuccessMetricType {
  ACCURACY = 'accuracy',
  PRECISION = 'precision',
  RECALL = 'recall',
  F1_SCORE = 'f1_score',
  LATENCY = 'latency',
  THROUGHPUT = 'throughput',
  CONVERSION_RATE = 'conversion_rate',
  REVENUE = 'revenue',
  CUSTOM = 'custom',
}

export enum CanaryPhase {
  INITIALIZING = 'initializing',
  TRAFFIC_SHIFT = 'traffic_shift',
  MONITORING = 'monitoring',
  PROMOTED = 'promoted',
  ROLLED_BACK = 'rolled_back',
  FAILED = 'failed',
}

export enum CanaryStrategy {
  LINEAR = 'linear',
  EXPONENTIAL = 'exponential',
  CUSTOM = 'custom',
}

// ============================================================================
// Model Serving Types
// ============================================================================

export interface ResourceRequirements {
  limits?: Record<string, string>;
  requests?: Record<string, string>;
}

export interface PredictorConfig {
  predictor_type: PredictorType;
  model_uri: string;
  runtime_version?: string;
  protocol?: string; // v1 or v2
  storage_uri?: string;
  framework?: string;
  device?: string; // cpu or gpu
  replicas?: number;
  resource_requirements?: ResourceRequirements;
  batch_size?: number;
  max_batch_size?: number;
  timeout?: number;
  custom_predictor_image?: string;
  custom_predictor_args?: string[];
  env?: Record<string, string>;
}

export interface ABTestConfig {
  experiment_id: string;
  model_variants: ModelVariant[];
  duration?: string;
  sample_size?: number;
  success_metric: string;
  success_mode: string;
  min_sample_size: number;
  traffic_split_method: string;
}

export interface ModelVariant {
  name: string;
  model_uri: string;
  model_version?: string;
  traffic_percentage: number;
  predictor_config?: PredictorConfig;
}

export interface CanaryConfig {
  canary_model_uri: string;
  canary_predictor_config: PredictorConfig;
  baseline_model_uri: string;
  baseline_predictor_config: PredictorConfig;
  canary_traffic_percentage: number;
  auto_promote: boolean;
  promotion_threshold: number;
  monitoring_window: string;
  min_requests: number;
  auto_rollback: boolean;
  rollback_threshold: number;
}

export interface InferenceService {
  name: string;
  namespace: string;
  description?: string;
  tags: string[];
  platform: ServingPlatform;
  mode: DeploymentMode;
  predictor_config?: PredictorConfig;
  ab_test_config?: ABTestConfig;
  canary_config?: CanaryConfig;
  endpoint?: string;
  url?: string;
  autoscaling_enabled: boolean;
  min_replicas: number;
  max_replicas: number;
  target_requests_per_second: number;
  enable_logging: boolean;
  log_url?: string;
  status: ServingStatus;
  status_message?: string;
  metadata: Record<string, unknown>;
  traffic_distribution?: Record<string, number>;
  project_id?: number;
  owner_id?: number;
  created_at: string;
  updated_at: string;
}

export interface ServiceMetrics {
  request_count: number;
  request_success_rate: number;
  avg_latency_ms: number;
  p50_latency_ms: number;
  p95_latency_ms: number;
  p99_latency_ms: number;
  throughput_per_second: number;
}

// ============================================================================
// A/B Testing Types
// ============================================================================

export interface VariantMetrics {
  variant_id: string;
  name: string;
  traffic_percentage: number;
  request_count: number;
  success_rate: number;
  avg_latency_ms: number;
  is_enabled: boolean;
  is_winner: boolean;
  custom_metrics: Record<string, number>;
}

export interface ABTestExperiment {
  experiment_id: string;
  name: string;
  description?: string;
  is_active: boolean;
  is_running: boolean;
  has_minimum_samples: boolean;
  success_metric: string;
  split_method: string;
  variants: VariantMetrics[];
  winner_variant_id?: string;
  project_id?: number;
  created_at: string;
  updated_at: string;
}

export interface StatisticalTestResult {
  is_significant: boolean;
  p_value: number;
  confidence_interval: [number, number];
  effect_size: number;
  control_metric: number;
  treatment_metric: number;
  relative_improvement: number;
  should_promote: boolean;
  recommendation: string;
}

export interface CreateABTestRequest {
  name: string;
  description?: string;
  variants: Omit<ModelVariant, 'predictor_config'>[];
  success_metric?: SuccessMetricType;
  split_method?: TrafficSplitMethod;
  duration_hours?: number;
  min_sample_size?: number;
  confidence_level?: number;
  project_id?: number;
  tags?: string[];
}

// ============================================================================
// Canary Deployment Types
// ============================================================================

export interface CanaryStep {
  step_number: number;
  traffic_percentage: number;
  duration_minutes: number;
  min_requests: number;
  max_error_rate?: number;
  max_latency_p95_ms?: number;
  status: 'pending' | 'in_progress' | 'passed' | 'failed';
  started_at?: string;
  completed_at?: string;
  actual_error_rate: number;
  actual_latency_p95_ms: number;
  total_requests: number;
}

export interface CanaryDeployment {
  deployment_id: string;
  name: string;
  service_name: string;
  phase: CanaryPhase;
  current_step: number;
  total_steps: number;
  current_traffic_percentage: number;
  progress_percentage: number;
  is_running: boolean;
  is_complete: boolean;
  status_message?: string;
  baseline_model: string;
  canary_model: string;
  steps: CanaryStep[];
  latest_metrics?: {
    canary_error_rate: number;
    canary_latency_p95_ms: number;
    canary_request_count: number;
  };
  created_at: string;
  started_at?: string;
  completed_at?: string;
}

export interface CreateCanaryRequest {
  service_name: string;
  baseline_model_uri: string;
  baseline_version?: string;
  canary_model_uri: string;
  canary_version?: string;
  strategy?: CanaryStrategy;
  steps?: number;
  duration_minutes?: number;
  auto_promote?: boolean;
  auto_rollback?: boolean;
  rollback_threshold?: number;
  max_error_rate?: number;
  max_latency_p95_ms?: number;
}

// ============================================================================
// Form Types
// ============================================================================

export interface CreateServiceFormData {
  name: string;
  namespace: string;
  description?: string;
  tags: string[];
  platform: ServingPlatform;
  mode: DeploymentMode;
  predictor_config?: PredictorConfig;
  autoscaling_enabled: boolean;
  min_replicas: number;
  max_replicas: number;
  target_requests_per_second: number;
  enable_logging: boolean;
}

export interface ServiceFilters {
  status?: ServingStatus;
  platform?: ServingPlatform;
  mode?: DeploymentMode;
  search?: string;
  project_id?: number;
}

// ============================================================================
// Presets & Constants
// ============================================================================

export const PLATFORM_OPTIONS = [
  { label: 'KServe', value: ServingPlatform.KSERVE, description: 'Kubernetes-native inference' },
  { label: 'Seldon Core', value: ServingPlatform.SELDON, description: 'Advanced deployment strategies' },
  { label: 'Triton', value: ServingPlatform.TRITON, description: 'NVIDIA Triton Inference Server' },
  { label: 'Custom', value: ServingPlatform.CUSTOM, description: 'Custom deployment' },
];

export const DEPLOYMENT_MODE_OPTIONS = [
  { label: 'Single Model', value: DeploymentMode.RAW, icon: '⚡', description: 'Deploy a single model' },
  { label: 'A/B Testing', value: DeploymentMode.AB_TESTING, icon: '🧪', description: 'Test multiple models with traffic split' },
  { label: 'Canary', value: DeploymentMode.CANARY, icon: '🐤', description: 'Gradual traffic shift' },
  { label: 'Shadow', value: DeploymentMode.SHADOW, icon: '👻', description: 'Shadow traffic to new model' },
  { label: 'Mirrored', value: DeploymentMode.MIRRORED, icon: '🪞', description: 'Multiple model mirrors' },
];

export const PREDICTOR_TYPE_OPTIONS = [
  { label: 'Scikit-learn', value: PredictorType.SKLEARN },
  { label: 'XGBoost', value: PredictorType.XGBOOST },
  { label: 'LightGBM', value: PredictorType.LIGHTGBM },
  { label: 'PyTorch', value: PredictorType.PYTORCH },
  { label: 'TensorFlow', value: PredictorType.TENSORFLOW },
  { label: 'ONNX', value: PredictorType.ONNX },
  { label: 'HuggingFace', value: PredictorType.HUGGINGFACE },
  { label: 'Custom', value: PredictorType.CUSTOM },
];

export const SUCCESS_METRIC_OPTIONS = [
  { label: 'Accuracy', value: SuccessMetricType.ACCURACY, mode: 'max' },
  { label: 'Precision', value: SuccessMetricType.PRECISION, mode: 'max' },
  { label: 'Recall', value: SuccessMetricType.RECALL, mode: 'max' },
  { label: 'F1 Score', value: SuccessMetricType.F1_SCORE, mode: 'max' },
  { label: 'Latency', value: SuccessMetricType.LATENCY, mode: 'min' },
  { label: 'Throughput', value: SuccessMetricType.THROUGHPUT, mode: 'max' },
  { label: 'Conversion Rate', value: SuccessMetricType.CONVERSION_RATE, mode: 'max' },
  { label: 'Revenue', value: SuccessMetricType.REVENUE, mode: 'max' },
];

export const TRAFFIC_SPLIT_METHOD_OPTIONS = [
  { label: 'Fixed Split', value: TrafficSplitMethod.FIXED, description: 'Fixed percentage distribution' },
  { label: 'Epsilon-Greedy', value: TrafficSplitMethod.EPSILON_GREEDY, description: 'Explore-exploit balance' },
  { label: 'Thompson Sampling', value: TrafficSplitMethod.THOMPSON_SAMPLING, description: 'Bayesian optimization' },
  { label: 'UCB1', value: TrafficSplitMethod.UCB1, description: 'Upper confidence bound' },
];

export const CANARY_STRATEGY_OPTIONS = [
  { label: 'Linear', value: CanaryStrategy.LINEAR, description: 'Equal step increments' },
  { label: 'Exponential', value: CanaryStrategy.EXPONENTIAL, description: 'Accelerated rollout' },
  { label: 'Custom', value: CanaryStrategy.CUSTOM, description: 'Custom schedule' },
];

// Resource presets
export const RESOURCE_PRESETS: Record<string, Partial<PredictorConfig>> = {
  'cpu-small': {
    replicas: 1,
    resource_requirements: {
      requests: { cpu: '500m', memory: '512Mi' },
      limits: { cpu: '1000m', memory: '1Gi' },
    },
  },
  'cpu-medium': {
    replicas: 2,
    resource_requirements: {
      requests: { cpu: '1000m', memory: '1Gi' },
      limits: { cpu: '2000m', memory: '2Gi' },
    },
  },
  'cpu-large': {
    replicas: 3,
    resource_requirements: {
      requests: { cpu: '2000m', memory: '2Gi' },
      limits: { cpu: '4000m', memory: '4Gi' },
    },
  },
  'gpu-single': {
    replicas: 1,
    device: 'gpu',
    resource_requirements: {
      requests: { cpu: '1000m', memory: '2Gi', 'nvidia.com/gpu': '1' },
      limits: { cpu: '2000m', memory: '4Gi', 'nvidia.com/gpu': '1' },
    },
  },
  'gpu-quad': {
    replicas: 1,
    device: 'gpu',
    resource_requirements: {
      requests: { cpu: '4000m', memory: '8Gi', 'nvidia.com/gpu': '4' },
      limits: { cpu: '8000m', memory: '16Gi', 'nvidia.com/gpu': '4' },
    },
  },
};

export const AUTOSCALING_PRESETS = {
  conservative: {
    min_replicas: 1,
    max_replicas: 3,
    target_requests_per_second: 10,
  },
  moderate: {
    min_replicas: 2,
    max_replicas: 5,
    target_requests_per_second: 20,
  },
  aggressive: {
    min_replicas: 2,
    max_replicas: 10,
    target_requests_per_second: 50,
  },
};

// ============================================================================
// API Request/Response Types
// ============================================================================

export interface CreateServiceRequest {
  name: string;
  namespace?: string;
  description?: string;
  tags?: string[];
  platform?: ServingPlatform;
  mode?: DeploymentMode;
  predictor_config?: PredictorConfig;
  ab_test_config?: ABTestConfig;
  canary_config?: CanaryConfig;
  autoscaling_enabled?: boolean;
  min_replicas?: number;
  max_replicas?: number;
  target_requests_per_second?: number;
  enable_logging?: boolean;
  log_url?: string;
  metadata?: Record<string, unknown>;
  project_id?: number;
}

export interface UpdateTrafficRequest {
  traffic_distribution: Record<string, number>;
}

export interface ScaleServiceRequest {
  replicas: number;
}

File: apps/backend/app/services/serving/serving.py
"""
Model Serving Service

Provides model serving capabilities with KServe/Seldon integration
for deploying and managing inference services.
"""

import logging
import asyncio
import uuid
from typing import Dict, List, Optional, Any, Callable
from dataclasses import dataclass, field
from enum import Enum
from datetime import datetime, timedelta
import json

logger = logging.getLogger(__name__)


class ServingPlatform(str, Enum):
    """Model serving platforms"""

    KSERVE = "kserve"
    SELDON = "seldon"
    TRITON = "triton"
    CUSTOM = "custom"


class ServingStatus(str, Enum):
    """Serving service status"""

    PENDING = "pending"
    DEPLOYING = "deploying"
    RUNNING = "running"
    UPDATING = "updating"
    FAILED = "failed"
    STOPPED = "stopped"
    UNKNOWN = "unknown"


class PredictorType(str, Enum):
    """Predictor types"""

    SKLEARN = "sklearn"
    XGBOOST = "xgboost"
    LIGHTGBM = "lightgbm"
    PYTORCH = "pytorch"
    TENSORFLOW = "tensorflow"
    ONNX = "onnx"
    HUGGINGFACE = "huggingface"
    CUSTOM = "custom"


class DeploymentMode(str, Enum):
    """Deployment modes"""

    RAW = "raw"  # Single model deployment
    A_B_TESTING = "ab_testing"  # A/B testing
    CANARY = "canary"  # Canary deployment
    SHADOW = "shadow"  # Shadow deployment
    MIRRORED = "mirrored"  # Multiple model mirrors


@dataclass
class PredictorConfig:
    """Predictor configuration"""

    predictor_type: PredictorType
    model_uri: str
    runtime_version: Optional[str] = None
    protocol: str = "v1"  # v1 or v2
    storage_uri: Optional[str] = None

    # Model-specific settings
    framework: Optional[str] = None  # pytorch, tensorflow, xgboost, etc.
    device: str = "cpu"  # cpu or gpu

    # Resource requirements
    replicas: int = 1
    resource_requirements: Dict[str, str] = field(default_factory=dict)

    # Inference settings
    batch_size: Optional[int] = None
    max_batch_size: Optional[int] = None
    timeout: int = 60

    # Custom predictor
    custom_predictor_image: Optional[str] = None
    custom_predictor_args: List[str] = field(default_factory=list)

    # Environment variables
    env: Dict[str, str] = field(default_factory=dict)

    def to_kserve_inference_service(self) -> Dict[str, Any]:
        """Convert to KServe InferenceService spec"""
        spec = {
            "apiVersion": "serving.kserve.io/v1beta1",
            "kind": "InferenceService",
            "metadata": {
                "name": f"{self.predictor_type}-predictor",
            },
            "spec": {
                "predictor": {
                    "modelFormat": self.framework,
                    "protocol": self.protocol,
                },
            },
        }

        # Add model URI
        if self.storage_uri:
            spec["spec"]["storageUri"] = self.storage_uri
        else:
            spec["spec"]["model"] = {
                "model": {
                    "name": self.model_uri,
                    "framework": self.framework,
                }
            }

        # Add runtime
        if self.runtime_version:
            spec["spec"]["predictor"]["runtime"] = self.runtime_version

        # Add replicas
        if self.replicas > 1:
            spec["spec"]["replicas"] = self.replicas

        # Add resource requirements
        if self.resource_requirements:
            spec["spec"]["resources"] = self.resource_requirements

        return spec


@dataclass
class ABTestConfig:
    """A/B testing configuration"""

    experiment_id: str
    model_variants: List[Dict[str, Any]]  # Each variant has model_uri, predictor_config, traffic_percentage
    success_metric: str  # e.g., "accuracy", "conversion_rate"

    # Optional parameters with defaults
    duration: Optional[str] = None  # e.g., "7d", "24h"
    sample_size: Optional[int] = None
    success_mode: str = "max"  # "max" or "min"
    min_sample_size: int = 100
    traffic_split_method: str = "fixed"  # "fixed", "epsilon_greedy", "thompson_sampling"


@dataclass
class CanaryConfig:
    """Canary deployment configuration"""

    canary_model_uri: str
    canary_predictor_config: PredictorConfig

    baseline_model_uri: str
    baseline_predictor_config: PredictorConfig

    # Canary strategy
    canary_traffic_percentage: int = 10  # Start with 10% traffic
    auto_promote: bool = True
    promotion_threshold: float = 0.95  # Promote if canary is 5% better

    # Monitoring
    monitoring_window: str = "1h"  # Time window for metrics comparison
    min_requests: int = 100  # Minimum requests before evaluation

    # Rollback
    auto_rollback: bool = True
    rollback_threshold: float = 0.90  # Rollback if degradation > 10%


@dataclass
class InferenceService:
    """Inference service configuration"""

    # Basic info
    name: str
    namespace: str = "default"
    description: Optional[str] = None
    tags: List[str] = field(default_factory=list)

    # Deployment
    platform: ServingPlatform = ServingPlatform.KSERVE
    mode: DeploymentMode = DeploymentMode.RAW

    # Primary predictor
    predictor_config: Optional[PredictorConfig] = None

    # A/B testing
    ab_test_config: Optional[ABTestConfig] = None

    # Canary deployment
    canary_config: Optional[CanaryConfig] = None

    # Service settings
    endpoint: Optional[str] = None
    url: Optional[str] = None

    # Autoscaling
    autoscaling_enabled: bool = False
    min_replicas: int = 1
    max_replicas: int = 3
    target_requests_per_second: int = 10

    # Monitoring
    enable_logging: bool = True
    log_url: Optional[str] = None

    # Status
    status: ServingStatus = ServingStatus.PENDING
    status_message: Optional[str] = None

    # Metadata
    metadata: Dict[str, Any] = field(default_factory=dict)

    # Timestamps
    created_at: datetime = field(default_factory=datetime.utcnow)
    updated_at: datetime = field(default_factory=datetime.utcnow)

    # Ownership
    owner_id: Optional[int] = None
    project_id: Optional[int] = None

    def get_traffic_distribution(self) -> Dict[str, int]:
        """Get current traffic distribution across models"""
        if self.mode == DeploymentMode.RAW:
            return {"default": 100}
        elif self.mode == DeploymentMode.A_B_TESTING and self.ab_test_config:
            distribution = {}
            for variant in self.ab_test_config.model_variants:
                distribution[variant.get("name", "default")] = variant.get(
                    "traffic_percentage", 100 // len(self.ab_test_config.model_variants)
                )
            return distribution
        elif self.mode == DeploymentMode.CANARY and self.canary_config:
            return {
                "canary": self.canary_config.canary_traffic_percentage,
                "baseline": 100 - self.canary_config.canary_traffic_percentage,
            }
        return {"default": 100}


class ModelServingService:
    """
    Model serving service for deploying and managing inference services

    Supports KServe, Seldon Core, and custom serving platforms.
    """

    def __init__(
        self,
        platform: ServingPlatform = ServingPlatform.KSERVE,
        kube_config: Optional[Dict[str, Any]] = None,
    ):
        """
        Initialize model serving service

        Args:
            platform: Serving platform (kserve, seldon, etc.)
            kube_config: Kubernetes configuration
        """
        self.platform = platform
        self.kube_config = kube_config or {}

        # Track active services
        self._services: Dict[str, InferenceService] = {}

    async def deploy_service(
        self,
        service: InferenceService,
    ) -> InferenceService:
        """
        Deploy a model serving service

        Args:
            service: Service configuration

        Returns:
            Deployed service with status
        """
        service.status = ServingStatus.DEPLOYING

        logger.info(f"Deploying service {service.name} using {self.platform}")

        # Generate deployment manifest
        manifest = self._generate_deployment_manifest(service)

        # Submit to Kubernetes
        await self._submit_to_kubernetes(manifest, service.namespace)

        service.status = ServingStatus.RUNNING
        service.endpoint = f"{service.name}.{service.namespace}.svc.cluster.local"

        # Track service
        self._services[service.name] = service

        return service

    async def update_service(
        self,
        service_name: str,
        updated_config: InferenceService,
    ) -> InferenceService:
        """
        Update an existing service

        Args:
            service_name: Service name
            updated_config: New configuration

        Returns:
            Updated service
        """
        service = self._services.get(service_name)
        if not service:
            raise ValueError(f"Service {service_name} not found")

        # Handle mode transitions
        if service.mode != updated_config.mode:
            # Need to delete and recreate for mode changes
            await self.undeploy_service(service_name)
            updated_config.status = ServingStatus.DEPLOYING
            return await self.deploy_service(updated_config)

        # Update in-place
        service.status = ServingStatus.UPDATING
        service.predictor_config = updated_config.predictor_config
        service.ab_test_config = updated_config.ab_test_config
        service.canary_config = updated_config.canary_config

        # Generate updated manifest and apply
        manifest = self._generate_deployment_manifest(service)
        await self._apply_kubernetes_manifest(manifest, service.namespace)

        service.status = ServingStatus.RUNNING
        service.updated_at = datetime.utcnow()

        return service

    async def undeploy_service(
        self,
        service_name: str,
        namespace: Optional[str] = None,
    ) -> bool:
        """
        Undeploy a model serving service

        Args:
            service_name: Service name
            namespace: Kubernetes namespace

        Returns:
            True if undeployed successfully
        """
        service = self._services.get(service_name)
        namespace = namespace or (service.namespace if service else "default")

        if service:
            service.status = ServingStatus.STOPPED

        # Delete from Kubernetes
        await self._delete_from_kubernetes(service_name, namespace)

        # Remove from tracking
        if service_name in self._services:
            del self._services[service_name]

        return True

    async def get_service_status(
        self,
        service_name: str,
    ) -> Optional[ServingStatus]:
        """
        Get current service status

        Args:
            service_name: Service name

        Returns:
            Current status or None
        """
        # Query Kubernetes for service status
        return self._query_kubernetes_status(service_name)

    async def scale_service(
        self,
        service_name: str,
        replicas: int,
    ) -> bool:
        """
        Scale service replicas

        Args:
            service_name: Service name
            replicas: Number of replicas

        Returns:
            True if scaled successfully
        """
        service = self._services.get(service_name)
        if not service:
            return False

        # Update replicas in Kubernetes
        await self._scale_kubernetes_deployment(service_name, replicas, service.namespace)

        if service.predictor_config:
            service.predictor_config.replicas = replicas

        return True

    def get_traffic_distribution(
        self,
        service_name: str,
    ) -> Dict[str, int]:
        """
        Get current traffic distribution

        Args:
            service_name: Service name

        Returns:
            Traffic distribution percentages
        """
        service = self._services.get(service_name)
        if not service:
            raise ValueError(f"Service {service_name} not found")

        return service.get_traffic_distribution()

    async def update_traffic_split(
        self,
        service_name: str,
        traffic_distribution: Dict[str, int],
    ) -> bool:
        """
        Update traffic split for A/B testing or canary

        Args:
            service_name: Service name
            traffic_distribution: Dictionary mapping variant name to percentage

        Returns:
            True if updated successfully
        """
        service = self._services.get(service_name)
        if not service:
            return False

        # Validate percentages sum to 100
        if sum(traffic_distribution.values()) != 100:
            raise ValueError("Traffic percentages must sum to 100")

        # Update service configuration
        if service.mode == DeploymentMode.A_B_TESTING:
            for i, variant in enumerate(service.ab_test_config.model_variants):
                variant_name = variant.get("name", f"variant_{i}")
                if variant_name in traffic_distribution:
                    variant["traffic_percentage"] = traffic_distribution[variant_name]
        elif service.mode == DeploymentMode.CANARY:
            if "canary" in traffic_distribution:
                service.canary_config.canary_traffic_percentage = traffic_distribution["canary"]

        # Apply traffic split in Kubernetes
        await self._apply_traffic_split(service_name, traffic_distribution, service.namespace)

        return True

    async def get_service_metrics(
        self,
        service_name: str,
        duration: str = "1h",
    ) -> Dict[str, Any]:
        """
        Get service metrics

        Args:
            service_name: Service name
            duration: Time window for metrics

        Returns:
            Service metrics
        """
        # Query metrics from Prometheus/monitoring
        return await self._query_service_metrics(service_name, duration)

    def _generate_deployment_manifest(self, service: InferenceService) -> Dict[str, Any]:
        """Generate Kubernetes deployment manifest"""
        if service.platform == ServingPlatform.KSERVE:
            return self._generate_kserve_manifest(service)
        elif service.platform == ServingPlatform.SELDON:
            return self._generate_seldon_manifest(service)
        else:
            return self._generate_custom_manifest(service)

    def _generate_kserve_manifest(self, service: InferenceService) -> Dict[str, Any]:
        """Generate KServe InferenceService manifest"""
        manifest = {
            "apiVersion": "serving.kserve.io/v1beta1",
            "kind": "InferenceService",
            "metadata": {
                "name": service.name,
                "namespace": service.namespace,
                "annotations": service.metadata.get("annotations", {}),
                "labels": service.metadata.get("labels", {}),
            },
            "spec": {
                "predictor": {},
            },
        }

        # Add predictor configuration
        if service.predictor_config:
            predictor_spec = service.predictor_config.to_kserve_inference_service()
            manifest["spec"]["predictor"] = predictor_spec["spec"]["predictor"]

        # Add autoscaling
        if service.autoscaling_enabled:
            manifest["spec"].update({
                "autoscalingPolicy": {
                    "minReplicas": service.min_replicas,
                    "maxReplicas": service.max_replicas,
                    "target": {
                        "requestsPerSecond": service.target_requests_per_second,
                    },
                },
            })

        return manifest

    def _generate_seldon_manifest(self, service: InferenceService) -> Dict[str, Any]:
        """Generate Seldon Deployment manifest"""
        # Generate SeldonDeployment manifest
        return {
            "apiVersion": "machinelearning.seldon.io/v1",
            "kind": "SeldonDeployment",
            "metadata": {
                "name": service.name,
                "namespace": service.namespace,
            },
            "spec": {
                "predictors": [{
                    "componentSpec": {
                        "spec": {
                            "containers": [{
                                "name": "predictor",
                                "image": service.predictor_config.custom_predictor_image,
                                "resources": service.predictor_config.resource_requirements,
                            }],
                        },
                    },
                }],
            },
        }

    def _generate_custom_manifest(self, service: InferenceService) -> Dict[str, Any]:
        """Generate custom deployment manifest"""
        # Generate Kubernetes Deployment and Service
        return {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "name": service.name,
                "namespace": service.namespace,
            },
            "spec": {
                "replicas": service.predictor_config.replicas if service.predictor_config else 1,
                "selector": {
                    "app": service.name,
                },
                "template": {
                    "metadata": {
                        "labels": {"app": service.name},
                    },
                    "spec": {
                        "containers": [{
                            "name": "predictor",
                            "image": service.predictor_config.custom_predictor_image,
                            "resources": service.predictor_config.resource_requirements,
                            "ports": [{"containerPort": 8080}],
                            "env": [{"name": k, "value": v} for k, v in service.predictor_config.env.items()],
                        }],
                    },
                },
            },
        }

    async def _submit_to_kubernetes(
        self,
        manifest: Dict[str, Any],
        namespace: str,
    ):
        """Submit manifest to Kubernetes"""
        # In production, use kubernetes-python client
        logger.info(f"Submitting to Kubernetes: {manifest['kind']} - {manifest['metadata']['name']}")
        # await self._kube_client.create_namespaced_custom_object(...)
        await asyncio.sleep(0.1)  # Placeholder

    async def _apply_kubernetes_manifest(
        self,
        manifest: Dict[str, Any],
        namespace: str,
    ):
        """Apply manifest to Kubernetes"""
        logger.info(f"Applying to Kubernetes: {manifest['kind']} - {manifest['metadata']['name']}")
        # await self._kube_client.patch_namespaced_custom_object(...)
        await asyncio.sleep(0.1)  # Placeholder

    async def _delete_from_kubernetes(self, name: str, namespace: str):
        """Delete resource from Kubernetes"""
        logger.info(f"Deleting from Kubernetes: {name} in {namespace}")
        # await self._kube_client.delete_namespaced_custom_object(...)
        await asyncio.sleep(0.1)  # Placeholder

    async def _scale_kubernetes_deployment(
        self,
        name: str,
        replicas: int,
        namespace: str,
    ):
        """Scale deployment in Kubernetes"""
        logger.info(f"Scaling {name} to {replicas} replicas")
        # await self._kube_client.patch_namespaced_deployment_scale(...)
        await asyncio.sleep(0.1)  # Placeholder

    async def _query_kubernetes_status(self, service_name: str) -> ServingStatus:
        """Query service status from Kubernetes"""
        # In production, query actual service status
        return ServingStatus.RUNNING

    async def _apply_traffic_split(
        self,
        service_name: str,
        traffic_distribution: Dict[str, int],
        namespace: str,
    ):
        """Apply traffic split configuration"""
        logger.info(f"Applying traffic split for {service_name}: {traffic_distribution}")
        # Update Istio VirtualService or equivalent
        await asyncio.sleep(0.1)  # Placeholder

    async def _query_service_metrics(
        self,
        service_name: str,
        duration: str,
    ) -> Dict[str, Any]:
        """Query service metrics from monitoring system"""
        # Return mock metrics for now
        return {
            "request_count": 10000,
            "request_success_rate": 0.995,
            "avg_latency_ms": 45.2,
            "p50_latency_ms": 32,
            "p95_latency_ms": 78,
            "p99_latency_ms": 145,
            "throughput_per_second": 220,
        }


# Singleton instance
_serving_service: Optional[ModelServingService] = None


def get_serving_service(
    platform: ServingPlatform = ServingPlatform.KSERVE,
    kube_config: Optional[Dict[str, Any]] = None,
) -> ModelServingService:
    """Get or create serving service singleton"""
    global _serving_service
    if _serving_service is None or _serving_service.platform != platform:
        _serving_service = ModelServingService(platform, kube_config)
    return _serving_service
